Devscord.DiscordFramework/Framework/Commands/Services/BotCommandsMatchingService.cs

```diff
@@ -1,87 +1,146 @@
-﻿using Devscord.DiscordFramework.Framework.Commands.Parsing.Models;
+﻿using Devscord.DiscordFramework.Commons.Exceptions;
+using Devscord.DiscordFramework.Framework.Commands.Parsing.Models;
 using Devscord.DiscordFramework.Framework.Commands.Properties;
+using Serilog;
+using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Text;
 using System.Text.RegularExpressions;
 
 namespace Devscord.DiscordFramework.Framework.Commands.Services
 {
     public class BotCommandsMatchingService
     {
+        private readonly Regex _exTime = new Regex(@"\d+(ms|d|h|m|s)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+        private readonly Regex _exUserMention = new Regex(@"\<@!?\d+\>", RegexOptions.Compiled);
+        private readonly Regex _exChannelMention = new Regex(@"\<#\d+\>", RegexOptions.Compiled);
 
-        private readonly Regex _exTime = new Regex(@"\d+(h|m|s)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
-        private readonly Regex _exUserMention = new Regex(@"\<@[^\d]*\d+\>", RegexOptions.Compiled);
-        private readonly Regex _exChannelMention = new Regex(@"\<#[^\d]*\d+\>", RegexOptions.Compiled);
-
-        public bool IsMatchedWithCommand(DiscordRequest request, BotCommandTemplate template)
+        public bool IsDefaultCommand(BotCommandTemplate template, IEnumerable<DiscordRequestArgument> arguments, bool isCommandMatchedWithCustom)
         {
-            if (!request.IsCommandForBot)
+            var requiredProperties = template.Properties.Where(x => !x.IsOptional);
+            if (requiredProperties.Any(property => arguments.All(arg => property.Name.ToLowerInvariant() != arg.Name?.ToLowerInvariant())))
             {
                 return false;
             }
-            if (request.Name.ToLowerInvariant() != template.NormalizedCommandName)
+            if (!isCommandMatchedWithCustom)
             {
-                return false;
+                return true;
             }
-            if (!this.CompareArgumentsToProperties(request.Arguments.ToList(), template.Properties.ToList()))
+            // if we give the command in such a way that isCommandMatchedWithCustom is true and it has been checked before that all the required arguments have been given and it turned out that they were given in this way, 
+            // the program checks whether all given arguments (even the optional ones) are known to the command, so it is known whether we are dealing with a command in the default form or in the form of custom.
+            // if the method returns false after the following operations, it means that the command is in a custom form
+            var argsWithoutListSubarguments = arguments
+                .Select(x => x.Name?.ToLowerInvariant())
+                .ToList();
+            var lists = template.Properties
+                .Where(property => property.Type == BotCommandPropertyType.List)
+                .Select(x => x.Name.ToLowerInvariant())
+                .ToList();
```

Comment:
```suggestion
                .ToHashSet();
```

Reply:
w jakim celu to ma być zamienione na HashSet?

Reply:
lekka poprawa wydajności
wygoogluj sobie porównanie szybkości metody Contains na List a HashSet
ogólnie w tym przypadku mało to zmienia, różnica będzie prawie niezauważalna
ale to jest zerowy koszt optymalizacji - wystarczy zmienić jedno słowo, a w żaden sposób nie ucierpi przez to czytelność kodu

Reply:
z drugiej strony HashSet tworzy sie dłużej niż lista
bo dodatkowo dla każdego elementu masz operacje sprawdzenia czy on już istnieje
a dodatkowo tworzenie dla niego hasha, żeby w przyszłości szybciej porównywać z nim nowe obiekty

Reply:
HashSet jest szybszy podczas samego używania
ale jego tworzenie jest bardziej kosztowne

Reply:
a no w sumie
to niech zostanie ta lista